Improve monitoring information of WMAgent components

vkuznet opened the pull request:
Fixes #12145 

#### Status
ready

#### Description
To improve WMagent component's monitoring we introduced the following changes:

- Add new procfs module with `processStatus` function based on `/proc` info for given PID. This function returns information about process threads which later can be used by WMStats. For example, here is addition to [`agentInfo`](https://cmsweb-testbed.cern.ch/couchdb/wmstats/_design/WMStats/_view/agentInfo) dictionary we supply to CouchDB:
```
  "components": {
...
    "DBS3Upload": [
      {"process": "python", "pid": "1921054", "status": "S (sleeping)",  "type": "process"},
      {"process": "python", "pid": "1921059", "status": "S (sleeping)",  "type": "thread" }
    ],
...
}
```
It shows the main process and its threads, in this case main process pid is 1921054, and pid of its threads (in this case only one additional thread) is 1921059 which is in sleeping state.

- add `processThreadsInfo` function in `Utils/ProcessStats.py` module which provides process and thread monitoring information:
```
{"process_name": .., "pid": .., "status": .., "ppid": .., "cmdline":.., 
 "cpu_usage_percent": .., "memory_usage_percent": .., "memory_rss": .., 
 "num_open_files": .., "num_connections": .., 
 "threads": [..]}
```
and thread info:
```
{"thread_id": .., "user_time":.., "system_time":.., "cpu_usage_percent": ..,
  "memory_usage_bytes":.., "num_open_files":.., "num_connections":..,
  "state":.., "name":..}
```
- update `WMComponent/AnalyticsDataCollector/DataCollectAPI.py` parts with information about dead threads via `threadsDetails` function within `agentInfo['down_component_detail']` used by WMStats web UI
- update `bin/wmcoreD` where we provide status of component's process along with its threads, e.g.
```
manage status
...
Component:WorkQueueManager 3545240 running with threads [3545241, 3545242, 3545243, 3545244]
Component:DBS3Upload 3545249 running with threads [3545254]
...
```
Each component now reports all its threads and if thread is missing/died/stuck it will provide proper message about it, e.g. component can be in `partially-running` state where its main process and some of the threads are ok but other threads misbehave

- introduced `trheads.json` file in component area along with `Daemon.xml` to capture initial conditions of component state. This information is used by aforementioned new functionality to determine if original threads are running or died

- propagate thread information into `WMComponent/AnalyticsDataCollector/DataCollectAPI.py` who use it and analyze state of specific component and properly report its state (with new threads information) to CouchDB.

**Note:** The information reported in WMStats web UI **agentInfo** tab can be accessed via the following URL: `https://xxxx.cern.ch/couchdb/wmstats/_design/WMStats/_view/agentInfo` which return nested dictionary which is used by WMStats JavaScript engine to report this information in **agentInfo** tab.

**Pylint note:** I noticed degradation in modified `bin/wmcoreD` codebase which is not related to proposed changes. For instance, there are underlined variables used in this codebase. And, I explicitly did not touch those errors/warnings which most likely come from stricter rules imposed by pylint over time. If those should be addressed I suggest to make an explicit request for them and I'll be happy to fix those issues.

**Note II:** The proposed changed to `wmcoreD` will affect output of `manage status` for list of components. In particular the new output for each individual component will have the following outputs:
- if everything is fine
```
Component:ABC 123 running with threads X
```
- if main process is dead
```
Component:ABC 123 not-running
```

- if main process is alive but one of its threads is dead
```
Component:ABC 123 partially-running with threads X, lost Y threads
```


#### Is it backward compatible (if not, which system it affects?)
YES

#### Related PRs

#### External dependencies / deployment changes

vkuznet committed:
add thread info status to WMAgent info pushed to CouchDB

vkuznet committed:
unit tests changes

amaltaro reviewed (commented):
Without having a deep look into the details of the worker threads implementation, but seeing each component worker thread inherits a `BaseWorkerThread`:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WorkerThreads/BaseWorkerThread.py

and runs as a daemon, spawning a Daemon.xml file, e.g.:
```
(WMAgent-2.3.9.2) [xxx@vocms0xxx:current]$ cat install/WorkQueueManager/Daemon.xml
```

Can't we consolidate most of this development into this module:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/Agent/Daemon/Details.py
?

Or perhaps we could have the relevant worker thread monitoring logic in the BaseWorkerThread itself:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WorkerThreads/BaseWorkerThread.py

In addition, I see the Harness module is used for starting up a component:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/Agent/Harness.py
but I think it is a higher layer and would not fit well for the worker threads monitoring.

vkuznet commented:
> src/python/WMCore/WorkerThreads/BaseWorkerThread.py

Alan, please go through the review to avoid triggering unnecessary work. In my view, what you suggest does not fit in existing architectural design because we don't have ability to communicate with threads from external process. In other words, our components and pollers are python processes rather (HTTP) services where you can request some monitoring metrics. In external process with thread model what can be done is to push metrics somewhere, e.g. we can push such information to internal database, or Monit infrastructure and then use them afterwards. At the moment we rely on process ID (extracted from Daemon.xml) to inspect its status about run-time environment of components/daemons.

I provided a stand-along functions for that and they are used in `Details.py` as you mentioned in your comment. Therefore, I recommend that you review the current changes, and if you have specific idea how to implement thread monitoring via integration with `[WMCore/WorkerThreads/BaseWorkerThread.py](https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WorkerThreads/BaseWorkerThread.py)` I suggest that you outline your idea as proposal for such monitoring. For instance
- add process metrics to BaseWorkerThread
- how to push/collect these metrics elsewhere (please specify where)
- how to extract these metrics from elsewhere and use it for monitoring of running components/daemons. For instance:
  - how to use these metrics in `manage status`
  - how to use these metrics in `WMComponent/AnalyticsDataCollector/DataCollectAPI.py` which report them to CouchDB, i.e. how it can communicate with run-time environment (`BaseWorkerThread.py`)
  - maybe we can push metrics directly from `BaseWorkerThread.py` to CouchDB, is it desired?

My point is that without global picture with all details of monitoring metrics workflows it is very hard to understand what you have in mind, and simply mentioning different parts of WMCore does not help to understand if it is feasible to accomplish with current architectural design.

amaltaro commented:
@vkuznet I am puzzled with the fragmentation of process-related code. Despite the extensive PR description (thanks!), please help me understand the following:
a) what is `wmcoreD` bin used for? I guess it is used by the `manage` script, hence the fragmentation from the components codebase
b) can't we merge the functionality in the new module `ProcFS` into `ProcessStats`? Is there any reason to keep those separated?
c) instead of extending `Utilities` module, can't we use the `Daemon/Details` module - which already loads the Daemon.xml and gives you the ProcessID element?
d) in `DataCollectAPI`, instead of relying in a broken information from the database - and the new data struct created - don't you think it is better to persist the actual worker thread ID in the database (wm_workers table?)

Checking the information already avaialble in WMStats `agentInfo` view, I see:
```
        "workers": [
          {
            "name": "AgentStatusPoller",
            "last_updated": 1743074171,
            "state": "Running",
            "poll_interval": 300,
            "cycle_time": 1.5446
          },
          {
            "name": "AnalyticsPoller",
            "last_updated": 1743074124,
            "state": "Running",
            "poll_interval": 600,
            "cycle_time": 0.8201
          },
...
```

and I would be in favor of always reporting the state of all worker threads (+ main component), instead of just publishing when one of the threads is down.

What do you think?

mapellidario commented:
Before I leave a review, I have a curiosity, mainly for my education, and if it has already been discussed i am sorry if i missed it. Is it there any reason why you use read the content of the `/proc/$pid` files in some places and use the python module `psutils` somewhere else? do they provide different information? if `psutil` can provide all the information we need, i would be in favor of never reading directly from `/prod/$pid` files

amaltaro commented:
Thank you for these clarifications, Valentin.

> b) we can merge ProcFS into ProcessStats but they have code not related to each other. The former keeps code to work with Linux file system, while later the code to about process monitoring. Because of this reason I kept them separately but I don't mind to merge them.

Given that we are still checking information for processes, I would merge them all together (probably keeping ProcessStats module). 

> c) the only reason I added extractFromXML due to its generic nature, i.e. this function can be used with any XML parsing task. Moving this from Utilities to Daemon/Details is acceptable but based on my previous reviews you always wanted to keep pieces in right places, and to me keeping generic function in Utilities is more appropriate then in Daemon/Details. And, I do use processStatus function I provided in Daemon/Details. If my answers are not satisfactory to you please elaborate more how do you prefer relocate the code. In my view everything is in right places and used accordingly.

Yes, generic functions are better placed under the `Utilities` package. However, in this case you are dealing with an agent component/daemon, which creates an XML. From what I can see, the Details.py module:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/Agent/Daemon/Details.py

exist exactly to deal with that Daemon.xml file. So it is somehow natural to me that we would use that module to retrieve the node that you are looking for. If not yet available, then we can extend that class with the required method.

Generic module does the job and can be used by anything else that needs to parse an xml. The inconsistency accessing Daemon.xml is what bugs me. Just my 2 cents, stick to whatever you/people prefer.

> d) I don't really have a clue why we keep process information in a database as it is outdated due to different time it is used and collected. ...

Data in the database will only be stale if a process dies, which is actually our clue to see that that component/process is no longer running.
The way the agent works is that, whenever we start a new component, it self-registers in the database (plus some information). If the row was already populated with an outdated PID, it is updated as well. So that table is always giving us a summary of components / worker threads / process id (well, without the thread id, which is being resolved with this PR). I am adding a couple of database screenshots for your consideration as well.

> @mapellidario , answering your question. Even though psutil is well known python package the Linux /proc file system is foundation of everything in Linux OS. I trust it more :) That said, I provided both ways, either use ProcFS module based on /proc FS, or use psutil approach in ProcessStats module.

It looks like `psutil` relies on `/proc` and `/sys` for linux process management, so we are probably fine if we adopt only one solution. I'd stick with one well tested solution, than having 2 and not being sure which one to take.

<img width="435" alt="wm_components" src="https://github.com/user-attachments/assets/aa9967ae-1727-429f-a849-4f606ed341ad" />
<img width="1076" alt="wm_workers" src="https://github.com/user-attachments/assets/7343648e-f0ea-4976-80cd-c341a560862c" />

vkuznet commented:
> Data in the database will only be stale if a process dies, which is actually our clue to see that that component/process is no longer running.

The data in the database will not be correct if **ANY** thread will die with main process still alive. For instance, if main process is fine and all of its threads too then your statement is correct. But if main process is fine but any of its threads are not (it is possible that thread dies but main process is alive) then content of database is not representing actual state of the component. For that reason I think we cannot rely on database content for any component (in this regard I don't know why do we need the database table for them either).

Before making any changes I want you to review this statement and decide how to proceed. In my view the database entries are not useful to represent state of the component and we **must** use either proc fs or pstuils to determine state of all component threads each time of polling cycle.

I'm tagging everyone to make awareness of this situation: @amaltaro , @anpicci , @todor-ivanov , @mapellidario , @khurtado , @d-ylee

vkuznet commented:
@amaltaro , thanks for explaining the algorithm but this is not what is implemented in a code. According to
- [isAlive](https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/Agent/Daemon/Details.py#L84) function of Daemon/Details.py module it checks 
```
        # Reference: ps -T -p 1946167 -o euser,pid,ppid,lwp,nlwp,stat,start
        # it prints the user, process and its threads, number of threads, etc
        dummyse, dummyso, rc = run('ps -p %s' % self['ProcessID'])
```
Please note that comment about ps command is correct (it should extract info about main process along with threads) the actual action does not do that, i.e. it only looks-up main process
```
# here is lookup of PID of Component:WorkQueueManager in my agent
ps -p 3545240
    PID TTY          TIME CMD
3545240 ?        04:16:45 python

# while if we'll supply -T flag it will print out all threads
ps -T -p 3545240
    PID    SPID TTY          TIME CMD
3545240 3545240 ?        00:00:00 python
3545240 3545241 ?        00:14:15 python
3545240 3545242 ?        00:10:18 python
3545240 3545243 ?        00:02:28 python
3545240 3545244 ?        03:49:19 python
```
- moreover, the ps command only show running process but does not show its actual state, the process may be alive but stuck on somthing (I/O), i.e. we need to know the state of the process which is what I suggest in this PR
- the [kill](https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/Agent/Daemon/Details.py#L96) function does not kill threads either it kills main process and even though it is rare in UNIX/Linux but it is not impossible for the main process to be killed and its thread be alive in a Zombie state. This happens when a thread is stuck in a kernel-level uninterruptible sleep (D state) like waiting for I/O (disk/NFS/network). In this case even SIGKILL (signal 9) may not immediately remove them until the blocking operation completes. And, even more the default parameter to `kill` function is signal 15 which is graceful shutdown and not even signal 9. Therefore, even though current code may work in majority of the case it is not 100%.

Therefore, according to my findins keeping main PID in DB table and according to current code implementation we do not cover by 100% all the case to kill components along with its threads but we do cover majority of them as it is usually sufficient to use main PID to kill process with its threads. I suggest that you carefully review my findings and refresh your knowledge about UNIX/Linux processes and threads and their management. In my view, the current code implementation is not sufficient to cover all corner cases and we should re-evaluate it further.

amaltaro reviewed (changes requested):
Valentin, please find a few more comments along the code.
In addition, given the amount of information in the PR description, I feel like we could move/enhance all of that to our documentation, for instance: https://gitlab.cern.ch/dmwm/wmcore-docs/-/blob/master/docs/wmcore/WMAgent-monitoring.md?ref_type=heads

and keep only a clear and concise summary of the changes provided with this PR (pointing to the documentation for further details). For instance, mentioning in the PR description about the new threads.json file and the new process_info would be useful.

amaltaro commented:
Now replying on top of your last comment:

> moreover, the ps command only show running process but does not show its actual state, the process may be alive but stuck on somthing (I/O), i.e. we need to know the state of the process which is what I suggest in this PR

At this stage, I don't think we need/want to implement actions based on the state of the process. So having the state in our monitoring is probably what we should focus on this contribution.

> the [kill](https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/Agent/Daemon/Details.py#L96) function does not kill threads either it kills ...

I might have missed it, but we do not have any changes in this PR on how components/threads are started or terminated, do we? I would suggest not to make any changes on that aspect, as I don't recall having such issues while operating agents. If needed, this is something that we can address in the future.

> Therefore, according to my findins keeping main PID in DB table and according to current code implementation we do not cover by 100% all the case to kill components along with its threads but we do cover majority of them as it is usually sufficient to use main PID to kill process with its threads. In my view, the current code implementation is not sufficient to cover all corner cases and we should re-evaluate it further.

As said above, I feel like the focus of this contribution is to have a better monitoring of the components and worker threads. If we have to modify how they are started up and/or killed/terminated, I would suggest to deal with this in a different ticket.

About the database table, I believe it could be a replacement to the threads.json file, assuming that the database would only give us the PID and the worker thread name that we want to look at. However, I think the `manage` script would not be able to easily use that - making the effort to make the DB table consistent probably not worth it. If I am missing anything, please let me know.

amaltaro commented:
The following thought just occurred. Instead of all this extra work to have a proper monitoring of the worker threads, why don't we separate each worker thread in its own component/process? I see many benefits in doing so, in addition to an easy way to scale out component instances in the future.

amaltaro commented:
Right. There are several benefits in separating these worker threads, and the disadvantages is only that some objects won't be shared between a components worker thread - but honestly, I don't know how much we are using this at the moment.

Nonetheless, I would not consider web-based components at the moment, because there are also security concerns, depending on where these services would be running. Other than separating worker threads in their own component, I think there are no further requirements to be changed in order to deliver a better monitoring of the components/worker threads - so it is a good compromise within the scope of this ticket.

vkuznet commented:
@amaltaro , so what do you propose to do with this PR? Should we leave and merge it, should we change PR to separate all threads into individual processes, if so how? Should we open new issue for such activity. It would be nice if you'll provide concrete steps how to move forward with issue in question.

amaltaro commented:
Before deciding how to proceed here, I think we need to answer the following questions:
a) do we resolve the ill worker thread monitoring if we isolate each in their own component? I think so, but are we missing anything else important?
b) which parts of this current implementation in this PR we **must** keep (resolves an immediate problem)?
c) which parts of this current implementation in this PR we **can** keep (might be useful at some point, but does not resolve any immediate problem)?

What I am trying to see is how much extra work is the component separation and whether that separation would suffice or not.

vkuznet commented:
Alan, let me answer your question one by one:
a) separation of component threads into individual processes will allow to use their PIDs to monitor their state. Therefore, the answer is yes.
b) and c):I think this PR just generalize **any** process monitoring since it provides additional functionality to obtain process information along with its threads. Will the process have threads or not does not concern of this PR. That said, if we'll have any daemon we want to get info about its process and threads this PR is beneficial. If we will never have threaded daemons or apps then we may not need it. For instance, any HTTP service we have is threaded (as we use CherryPy threaded framework), and therefore having proposed changes is beneficial in my view.

Finally, the component separation into individual processes has nothing to do with this PR, and it should go into separate issue. Moreover, it will require additional thoughts on several items like:
- agreement on naming convention. For instance, when we'll separate WorkQueueManager into separate processes, how we are going to name them?
- we may need to think how to adjust start-up procedure since right now it has one set of processes and we will expand it further
- we may need to adjust `manage` script for WMAgent to handle new processes too
- we should decide if we need to adjust any changes in internal database tables since we'll have more processes
- etc.

anpicci commented:
Hi @vkuznet @amaltaro, thank you for the fruitful discussion. I am in favor of concluding this PR as is, without further major development, and opening an issue to collect the idea outlined in the last messages.
Closing the issue related to this PR will impact https://github.com/dmwm/WMCore/pull/12308, and this is another motivation to moving along with merge this PR sooner rather than later

vkuznet commented:
@amaltaro , @anpicci this PR is ready, all commits are squashed and I don't know if you still want to have another review or not. Just in case I request Alan's review again.

amaltaro reviewed (changes requested):
@vkuznet from the discussion we had this morning, I thought you had addressed all comments/reviews, but I still see several comments along the code. 

Can you please update them accordingly? Or follow up with any discussion still needed? FYI @anpicci

amaltaro commented:
Valentin, in addition to the previous review, please update this documentation:
https://gitlab.cern.ch/dmwm/wmcore-docs/-/blob/master/docs/wmcore/ReqMgr2-monitoring.md

with the new content that will start getting pushed to WMStats (for the agent monitoring). I believe none of this affects the documents posted to AMQ, but please clarify that as well if needed.

vkuznet commented on test/python/Utils_t/Utilities_t.py:
```diff
@@ -4,18 +4,38 @@
     unittest for Utilities functions
     """
 
+    def setUp(self):
```
resolved.

vkuznet commented on src/python/Utils/ProcFS.py:
```diff
@@ -0,0 +1,52 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+"""
+File       : ProcFS.py
```
I adjusted the module to provide both implementation and use psutil as default one.

amaltaro replied:
To me, it would make sense to not fork the process status check into yet another module, and instead I'd integrate these changes together with src/python/Utils/ProcessStats.py. 
That's just my opinion though, perhaps others will share their view. Otherwise, please do not modify anything and let us move on with this.

vkuznet replied:
Originally, the idea was to have dedicated module for handling `procfs` functionality. That's why I created separate `ProcFS.py` from `ProcessStats.py` module. Now, per request from reviewers we converge on using `psutils` even though in your another comment you mentioned that usage of `psutil` on worker node may fail. Therefore, here I see few approaches:
- rollback to procfs (since psutil is not present everywhere) and then we may keep it in separate module, or
- merge `ProcFS.py` into `ProcessStats.py` as they both use psutil and provide into about process
- merge `ProcFS.py` into `ProcessStats.py` as they contain code about process/thread pids statuses

Either approach is fine with me and I can do whatever you'll consider appropriate.

anpicci replied:
In my view, we can leave as is, we can live with two scripts for the time being

vkuznet commented on src/python/Utils/ProcessStats.py:
```diff
@@ -89,6 +89,88 @@ def threadStack():
         threads.append(tdict)
     return dict(threads=threads)
 
+def processThreadsInfo(pid):
```
done

vkuznet commented on bin/wmcoreD:
```diff
@@ -320,18 +338,61 @@ def status(configFile):
+            orphanThreads.append(thread.id)
+
+    # Output result
+    status = "running" if processRunning else "not-running"
```
these words `running` or `not-running` will appear once someone runs `manage` script which calls `wmcoreD`. The actual message is defined a few lines below like:

```
    status = "running" if processRunning else "not-running"
    msg = f"Component:{component} {pid} {status} with threads {runningThreads}"
    if len(orphanThreads) > 0:
        status = "partially-running" if processRunning else "not-running"
        msg = f"Component:{component} {pid} {status} with threads {runningThreads}, lost {orphanThreads} threads"
    print(msg)
```

vkuznet commented on bin/wmcoreD:
```diff
@@ -320,18 +338,61 @@ def status(configFile):
+    status = "running" if processRunning else "not-running"
+    msg = f"Component:{component} {pid} {status} with threads {runningThreads}"
+    if len(orphanThreads) > 0:
+        status = "partially-running" if processRunning else "not-running"
```
yes, it means process is running (not gone) but one of its thread is dead such that message we get from `manage status` would be
```
Component:ABC 123 partially-running with threads X, lost Y threads
```

vkuznet commented:
> @vkuznet from the discussion we had this morning, I thought you had addressed all comments/reviews, but I still see several comments along the code.
> 
> Can you please update them accordingly? Or follow up with any discussion still needed? FYI @anpicci

@amaltaro , some of the changes were not pushed and I apologize for that. But I also forgot to update the comments you provided. Now it is done, and changes are pushed upstream. I verified that Jenkins are also fine. Per your request I updated the description with new statuses, see **Note II**. Once I'll update documentation I'll post relevant PR. So far this PR is ready.

@mapellidario , @anpicci I would appreciate if you'll make another round of review as well.

amaltaro commented on bin/wmcoreD:
```diff
@@ -235,6 +238,16 @@ def startup(configFile):
+        cpath = os.path.join(compDir, "threads.json")
+        if os.path.exists(cpath):
+            os.remove(cpath)
+        cxml = os.path.join(compDir, "Daemon.xml")
```
This line is dup of L231 (other than the variable name).

vkuznet replied:
ok, done

amaltaro commented on bin/wmcoreD:
```diff
@@ -320,18 +337,63 @@ def status(configFile):
+    if not os.path.exists(daemonXml):
+        print("Component:%s Not Running" % component)
+        return
+    daemon = Details(daemonXml)
```
I would suggest to be consistent on the parsing of the XML file, at least within these changes. Here you use the long implemented class to parse the XML, but at the top of this module you use `extractFromXML()`.

vkuznet replied:
switched to use extractFromXML

amaltaro commented on src/python/Utils/ProcFS.py:
```diff
@@ -0,0 +1,108 @@
+import os
+
+# 3rd party modules
+import psutil
```
I fear that importing this module in the worker node might hit an import error exception (as we do not have psutils in the grid nodes). So you might want to add a try/except here.

vkuznet replied:
Added try/except and fallback to use `processStatusViaProcFS`. But obvious question here is why do we need psutil implementation if on worker node we don't have it. The initial implementation via procfs is free from this issue.

anpicci replied:
I know that is not ideal, but if the try/except properly handle the events where `psutils` cannot be loaded, we are fine with that. Sorry if this point was missed previously, eventually requiring additional dev effort

amaltaro reviewed (changes requested):
Other than one import that I would strongly suggest to update, it looks good to me Valentin. However, please see inline for further feedback.

vkuznet commented:
@amaltaro , I applied changes based on your review and squashed commits. But I raised a question about `psutil` (see in my comments). Feel free to review again and let me know how to proceed.

mapellidario reviewed (approved):
thanks valentin for the great effort, i approve this PR

anpicci reviewed (approved):
Thank you for taking care of these changes. I approve them since to me no further changes are needed, we can keep `ProcFS.py` and `ProcessStats.py` (as well as their implementation) as they are, and come back to this whether needed
